README.md
# Hypercode

`Hypercode` is a npm package for Node.js that allows you to consume the [Hyper APIs](https://docs.gethyper.ai) easily in your JavaScript or TypeScript projects with complete **type-safety**. `Hypercode` handles the complexity of context management and response formatting, allowing you to focus on creating dynamic and intelligent features that enhance the user experience. It is a friendly npm package that makes it easy to get live, structured LLM responses with custom contexts in useful formats like integers, booleans, strings, and more.

Create a free account today on [Hyper](https://app.gethyper.ai) to start building your own custom contexts, integrating them into your applications, generating API keys, and using them in your projects with `Hypercode`.

Find the npm package [here](https://www.npmjs.com/package/hypercode)

[![npm version](https://badge.fury.io/js/hypercode.svg)](https://www.npmjs.com/package/hypercode)

**There are 3 main components to Hypercode:**

- Context Management (get information about the created contexts, more methods are coming soon) - [Learn More](#context-management)
- Response Formatting (get a response in a specific format or data type) - [Learn More](#response-formatting---types-in-hypercode)
- Embeddings Search (perform nuanced searches across integrated third-party data sources and internal documents) - [Learn More](#embeddings-search)

**Here's a quick example of how you can use Hypercode to get a boolean response:**

```javascript
const { data: isEarthFlat } = await hyper.types.boolean('Is the earth flat?');

console.log(isEarthFlat); // false
```

You can also pass the information along with your queries in the form of `context`. Context represents bundles of live data with relevance to the query, ensuring the LLM is given all the information necessary to produce an accurate response. You can build context objects in the [Hyper app](https://app.gethyper.ai), then use them in Hypercode:

```javascript
const { data: productLaunch } = await hyper.types.string(
  'When is the product launch?',
  {
    contextId: 'product-roadmap-context-id',
  },
);

console.log(productLaunch); // "The product launch is on December 11, 2023"
```

> **_NOTE:_** A context object is a collection of resources made up of files, web pages, and data from integrations like Google Drive, Slack, and GitHub. When you link a resource to a Context in Hyper, we generate embeddings that stay automatically synced with changes to your data.

The combination of structure and context in Hypercode is a powerful tool for adding sophisticated natural language understanding to your apps. Instead of complex setups, you can simply write async code that directly leverages AI insights for real-time decision-making and content creation:

```javascript
async function prepareEmailCampaign() {
  const todayIsHoliday = await hyper.types.boolean('Is today a holiday?', {
    contextId: 'company-holidays-context-id',
  });

  const emailSubject = await hyper.types.string(
    todayIsHoliday
      ? 'Generate a catchy subject for a post-holiday campaign'
      : 'Generate a catchy subject for a regular workday campaign',
    {
      contextId: 'spring-promo-context-id',
    },
  );

  return {
    campaign: 'Spring Promo',
    recipientList: '/lists/spring-campaign.csv',
    emailTemplate: 'templates/spring-promo.html',
    emailSubject,
    sendDate: todayIsHoliday ? 'next business day' : 'today',
  };
}

prepareEmailCampaign().then((requestBody) => {
  // Make an HTTP request to send the email
});
```

## Getting Started

### Step 1: Install Hypercode

Start by installing the Hypercode package through your preferred package manager:

#### Using npm

```bash
npm i hypercode
```

or

```bash
npm install --save hypercode
```

#### Using Yarn

```bash
yarn add hypercode
```

#### Using pnpm

```bash
pnpm add hypercode
```

### Step 2: Set your Hyper API Key

> **_NOTE:_** You need to generate an API key from the [API Key Settings](https://app.gethyper.ai/settings/api-keys) page in the Hyper app.

In your `.env` file, set your Hyper API Key:

```bash
HYPER_API_KEY="your_api_key_here"
```

Make sure to replace `your_api_key_here` with your actual Hyper API key obtained from the Hyper app.

### Step 3: Import Hypercode into Your Project

Import Hypercode in your JavaScript or TypeScript file to start using it:

```javascript
import { Hyper } from 'hypercode';
import dotenv from 'dotenv';

dotenv.config();

const hyper = new Hyper(process.env.HYPER_API_KEY);

// Use the `hyper` object to make different function calls
```

### Step 4: Start Querying

Now you're ready to start querying language models with your own custom contexts created with Hypercode!

## Examples

**You can find different example codes in the [examples](https://github.com/gethyperai/hypercode/tree/main/examples) folder.**

## Context Management

Hypercode allows you to programmatically manage your contexts that are created inside the [Hyper App](https://app.gethyper.ai). You can get information about the created contexts, and more methods are coming soon. More information about contexts can be found [here](https://docs.gethyper.ai/context).

**Here's a quick example of how you can use Hypercode to get all the contexts:**

```javascript
const { data: contexts, error } = await hyper.contexts.all();

console.log(contexts); // [{ id: 'context-id-1', name: 'Context 1', created_at: '2023-10-28T04:28:13.971776+00:00' }, { id: 'context-id-2', name: 'Context 2', created_at: '2023-11-02T22:18:44.978052+00:00' }]
```

## Response Formatting - Types in Hypercode

Hypercode provides a variety of structured query types, allowing you to seamlessly integrate LLM responses into your applications. You can utilize context with `contextId` when you need the model to consider specific background information for generating a response. More information about `types` can be found [here](https://docs.gethyper.ai/types) Here's a breakdown of the types and how they can be used:

- **string**: Get a simple string answer.

  ```javascript
  const { data: color, error } = await hyper.types.string(
    "What's the color of the sky?",
  );

  console.log(color); // "blue"
  ```

You can pass the `contextId` as an optional parameter to all the `types` methods. The `contextId` is the id of the context that you want to use for the query. Here's one example with the `integer` method:

- **integer**: Get an integer answer.

  ```javascript
  const { data: numberOfPatentsFiled, error } = await hyper.types.integer(
    'How many patents has the company filed since its inception?',
    { contextId: 'company-history-context-id' },
  );

  console.log(numberOfPatentsFiled); // 50
  ```

- **float**: Get a floating-point number answer.

  ```javascript
  const { data: averageRevenueGrowth, error } = await hyper.types.float(
    "What has been the company's average revenue growth rate over the last five years?",
    { contextId: 'financial-report-context-id' },
  );

  console.log(averageRevenueGrowth); // 4.7
  ```

- **boolean**: Get a true or false answer.

  ```javascript
  const { data: canCatsSeeInTheDark, error } = await hyper.types.boolean(
    'Can cats see in the dark?',
  );

  console.log(canCatsSeeInTheDark); // true
  ```

**You can also get the result as an array of the above types. Here are the methods for that:**

- `stringArray`, `integerArray`, `floatArray`, `booleanArray` methods: Get an array of the respective type as the answer.

  ```javascript
  // Get an array of strings as the answer
  const { data } = await hyper.types.stringArray('List all department names', {
    contextId: 'company-structure-context-id',
  });

  console.log(data); // ['Human Resources', 'Finance', 'Research and Development', 'Sales', 'Customer Support']

  // Get an array of integers as the answer
  const { data } = await hyper.types.integerArray(
    'What is the headcount for each department?',
    { contextId: 'company-structure-context-id' },
  );

  // Get an array of floats as the answer
  const { data } = await hyper.types.floatArray(
    'What were the customer satisfaction ratings from the last survey?',
    { contextId: 'customer-reviews-context-id' },
  );

  // Get an array of booleans as the answer
  const { data } = await hyper.types.booleanArray(
    'Are services meeting performance targets?',
    { contextId: 'performance-reviews-context-id' },
  );
  ```

## Embeddings Search

Utilize our Embeddings Search API to perform nuanced searches across integrated third-party data sources and internal documents. Leverage the `contextId` to scope searches to specific business contexts for enhanced relevance. More information about `search` can be found [here](https://docs.gethyper.ai/search).

```javascript
const { data, error } = await hyper.search.execute('quarterly sales report', {
  contextId: 'company-sales-reports-context-id',
});

console.log(data);
```

## Wrapping Up

With Hypercode, integrating live, structured responses in specific `types` from language models into your application has never been easier. By handling the complexity of managing the `contexts` and response formatting, Hypercode empowers you to focus on creating dynamic and intelligent features that enhance the user experience. With the powerful `search` feature, developers can perform nuanced searchs across integrated third-party data sources and internal documents to get the most relevant results from the context.

Remember, the examples provided are just a starting point. The potential uses of Hypercode are limited only by your imagination and the needs of your application. Whether you're building smart assistants, data analysis tools, content generators, or any other AI-driven application, Hypercode is designed to streamline your workflow and bring the power of LLMs to your fingertips.

Create a free account on [Hyper](https://app.gethyper.ai) to start building your own custom contexts and integrating them into your applications today!

## Bugs and Features

See the [issues](https://github.com/gethyperai/hypercode/issues) for a list of proposed features and known issues. Feel free to raise new issues.

## License

Distributed under the MIT License. See [LICENSE](https://github.com/gethyperai/hypercode/blob/main/LICENSE) for more information. © [Hyper AI Inc.](https://gethyper.ai)


CHANGELOG.md
## [2.2.4](https://github.com/gethyperai/hypercode/compare/v2.2.3...v2.2.4) (2023-12-11)

## [2.2.3](https://github.com/gethyperai/hypercode/compare/v2.2.2...v2.2.3) (2023-12-07)


### Bug Fixes

* **contexts:** replace `hyper.contexts.list` with `hyper.contexts.all` ([7b76910](https://github.com/gethyperai/hypercode/commit/7b769108bab8a5b6f4b3d3ee21fa539803aef106))
* **search:** metadata type for search result ([14fced7](https://github.com/gethyperai/hypercode/commit/14fced79ebb544df73d1e53aeaa40273acfdf991))

## [2.2.2](https://github.com/gethyperai/hypercode/compare/v2.2.1...v2.2.2) (2023-11-17)

## [2.2.1](https://github.com/gethyperai/hypercode/compare/v2.2.0...v2.2.1) (2023-11-17)

# [2.2.0](https://github.com/gethyperai/hypercode/compare/v2.1.0...v2.2.0) (2023-11-17)


### Bug Fixes

* **types:** make ts-types more stronger for `Types` ([ee605df](https://github.com/gethyperai/hypercode/commit/ee605df8d9e28831bbe8119d9be66c3715ae2e05))


### Features

* **jsdoc:** add JSDoc comments for `Contexts` ([d4f0d3d](https://github.com/gethyperai/hypercode/commit/d4f0d3d810f5fdb8e50e202aab01d08617b78c2b))
* **jsdoc:** add JSDoc comments for `Hyper` ([d81feb4](https://github.com/gethyperai/hypercode/commit/d81feb48f8e300124ab87d1d6c49dddf21a24460))
* **jsdoc:** add JSDoc comments for `Search` ([606268c](https://github.com/gethyperai/hypercode/commit/606268c1907df1862f129c33f3a563d755f66129))
* **jsodc:** add JSDoc comments for `Types` ([1155bc8](https://github.com/gethyperai/hypercode/commit/1155bc8308cc1092d5d241ee15eada4f26dfc55c))
* **types:** add new methods for `Types` ([dc6a0c4](https://github.com/gethyperai/hypercode/commit/dc6a0c4ee4b9ab8494ddb9d17bb9efd3c9fa1c77))

# [2.1.0](https://github.com/gethyperai/hypercode/compare/v2.0.0...v2.1.0) (2023-11-14)


### Bug Fixes

* **ts:** modify types for response ([6508862](https://github.com/gethyperai/hypercode/commit/65088628835c35dd58d5f489735c17fffacc4688))


### Features

* **contexts:** add `Contexts` class and methods ([73b7e7c](https://github.com/gethyperai/hypercode/commit/73b7e7cc6137943b4fd64ea9ee9e672ac714a935))
* **search:** add `Search` class inside `Hyper` ([bc37dfc](https://github.com/gethyperai/hypercode/commit/bc37dfc74c342be3fd44af40aac5468c60f38d39))

# [2.0.0](https://github.com/gethyperai/hypercode/compare/v1.1.1...v2.0.0) (2023-11-13)


### Bug Fixes

* **types:** pass `contextId` as options ([a0969c4](https://github.com/gethyperai/hypercode/commit/a0969c40250d98328178f4aabefb5b9ac572041f))


### Features

* **types:** add `Types` class ([ce41753](https://github.com/gethyperai/hypercode/commit/ce41753c5e18282e87a6bfc3666db3324c584199))
* **types:** add more methods ([df85ed1](https://github.com/gethyperai/hypercode/commit/df85ed147a50278f3b74be51562d59a34480a3f5))


### BREAKING CHANGES

* **types:** New Hyper API and methods

## [1.1.1](https://github.com/gethyperai/hypercode/compare/v1.1.0...v1.1.1) (2023-11-10)


### Bug Fixes

* **ci:** add permissions ([86ae312](https://github.com/gethyperai/hypercode/commit/86ae312c547605baae63df70d76e75a0016e0ad0))

# [1.1.0](https://github.com/gethyperai/hypercode/compare/v1.0.2...v1.1.0) (2023-11-10)


### Features

* **version:** bump version ([7752fca](https://github.com/gethyperai/hypercode/commit/7752fca9c134b98bb0545b10bb562ef25e76bf29))

## [1.0.2](https://github.com/gethyperai/hypercode/compare/v1.0.1...v1.0.2) (2023-11-10)


### Bug Fixes

* **ci:** semantic-release ([7362a40](https://github.com/gethyperai/hypercode/commit/7362a40a5f5b606971aa248170f685e417163907))

## [1.0.1](https://github.com/gethyperai/hypercode/compare/v1.0.0...v1.0.1) (2023-11-10)


### Bug Fixes

* **npm:** registry issues ([70d1b9f](https://github.com/gethyperai/hypercode/commit/70d1b9fb8615cc8b65b6a7b20697a96cdf4ac925))

# 1.0.0 (2023-11-10)


### Features

* **format:** bundle to `esm` and `cjs` ([7011213](https://github.com/gethyperai/hypercode/commit/701121358676758a2eb5f8cb9f835d93fac6b800))


src/types/types.ts
import { Hyper } from '../hyper';
import type { AllowedTypes, MethodParamsOptions } from './@types';

export class Types {
  constructor(private readonly hyper: Hyper) {}

  private async makeRequest<T>({
    endpointType,
    query,
    contextId,
  }: {
    endpointType: AllowedTypes;
    query: string;
    contextId?: string;
  }): Promise<
    | {
        data: T;
        error: null;
      }
    | {
        data: null;
        error: string;
      }
  > {
    const endpoint = `/types/${endpointType}`;
    const payload = contextId
      ? { query: query, context_id: contextId }
      : { content: query };

    const res = await this.hyper.post<{ data: T }>({
      endpoint,
      body: payload,
    });

    if (res.data) return { data: res.data.data, error: null };

    return { data: null, error: res.error };
  }

  /**
   * More information: https://docs.gethyper.ai/types#string
   * @param {string} query - The query to be processed
   * @param {MethodParamsOptions} options - Optional parameters to be passed to the request, such as `contextId`
   * @returns The result of the query as a string data type
   * @example
   * const { data, error } = await hyper.types.string(
        'Who is the CEO of SpaceX?',
        { contextId: '123e4567-e89b-12d3-a456-426614174000' },
     );
   * console.log(data); // Elon Musk
   */
  async string(query: string, options?: MethodParamsOptions) {
    const data = await this.makeRequest<string>({
      endpointType: 'string',
      query,
      contextId: options?.contextId,
    });

    return data;
  }

  /**
   * More information: https://docs.gethyper.ai/types#integer
   * @param {string} query - The query to be processed
   * @param {MethodParamsOptions} options - Optional parameters to be passed to the request, such as `contextId`
   * @returns The result of the query as an integer data type
   * @example
   * const { data, error } = await hyper.types.integer(
        'How many planets are in the Solar System?',
        { contextId: '123e4567-e89b-12d3-a456-426614174000' },
     );
   * console.log(data); // 8
   */
  async integer(query: string, options?: MethodParamsOptions) {
    const data = await this.makeRequest<number>({
      endpointType: 'integer',
      query,
      contextId: options?.contextId,
    });

    return data;
  }

  /**
   * More information: https://docs.gethyper.ai/types#float
   * @param {string} query - The query to be processed
   * @param {MethodParamsOptions} options - Optional parameters to be passed to the request, such as `contextId`
   * @returns The result of the query as a float data type
   * @example
   * const { data, error } = await hyper.types.float(
        'How many billion years old is the universe?',
        { contextId: '123e4567-e89b-12d3-a456-426614174000' },
     );
   * console.log(data); // 13.8
   */
  async float(query: string, options?: MethodParamsOptions) {
    const data = await this.makeRequest<number>({
      endpointType: 'float',
      query,
      contextId: options?.contextId,
    });

    return data;
  }

  /**
   * More information: https://docs.gethyper.ai/types#boolean
   * @param {string} query - The query to be processed
   * @param {MethodParamsOptions} options - Optional parameters to be passed to the request, such as `contextId`
   * @returns The result of the query as a boolean data type
   * @example
   * const { data, error } = await hyper.types.boolean(
        'Can cats see in the dark?',
        { contextId: '123e4567-e89b-12d3-a456-426614174000' },
     );
   * console.log(data); // true
   */
  async boolean(query: string, options?: MethodParamsOptions) {
    const data = await this.makeRequest<boolean>({
      endpointType: 'boolean',
      query,
      contextId: options?.contextId,
    });

    return data;
  }

  /**
   * More information: https://docs.gethyper.ai/types#datetime
   * @param {string} query - The query to be processed
   * @param {MethodParamsOptions} options - Optional parameters to be passed to the request, such as `contextId`
   * @returns The result of the query as a datetime data type
   * @example
   * const { data, error } = await hyper.types.datetime(
        'What is the date of the Apollo 11 moon landing?',
        { contextId: '123e4567-e89b-12d3-a456-426614174000' },
     );
   * console.log(data); // '1969-07-20T20:17:00Z'
   */
  async datetime(query: string, options?: MethodParamsOptions) {
    const data = await this.makeRequest<string>({
      endpointType: 'datetime',
      query,
      contextId: options?.contextId,
    });

    return data;
  }

  /**
   * More information: https://docs.gethyper.ai/types#string-array
   * @param {string} query - The query to be processed
   * @param {MethodParamsOptions} options - Optional parameters to be passed to the request, such as `contextId`
   * @returns The result of the query as a array of strings
   * @example
   * const { data, error } = await hyper.types.stringArray(
        'List all department names',
        { contextId: '123e4567-e89b-12d3-a456-426614174000' },
     );
   * console.log(data); // ['Human Resources', 'Finance', 'Research and Development', 'Sales', 'Customer Support']
   */
  async stringArray(query: string, options?: MethodParamsOptions) {
    const data = await this.makeRequest<string[]>({
      endpointType: 'string_array',
      query,
      contextId: options?.contextId,
    });

    return data;
  }

  /**
   * More information: https://docs.gethyper.ai/types#integer-array
   * @param {string} query - The query to be processed
   * @param {MethodParamsOptions} options - Optional parameters to be passed to the request, such as `contextId`
   * @returns The result of the query as a array of integers
   * @example
   * const { data, error } = await hyper.types.integerArray(
        'What is the headcount for each department?',
        { contextId: '123e4567-e89b-12d3-a456-426614174000' },
     );
   * console.log(data); // [25, 40, 15, 50, 30]
   */
  async integerArray(query: string, options?: MethodParamsOptions) {
    const data = await this.makeRequest<number[]>({
      endpointType: 'integer_array',
      query,
      contextId: options?.contextId,
    });

    return data;
  }

  /**
   * More information: https://docs.gethyper.ai/types#float-array
   * @param {string} query - The query to be processed
   * @param {MethodParamsOptions} options - Optional parameters to be passed to the request, such as `contextId`
   * @returns The result of the query as a array of floats
   * @example
   * const { data, error } = await hyper.types.floatArray(
        'What were the customer satisfaction ratings from the last survey?',
        { contextId: '123e4567-e89b-12d3-a456-426614174000' },
     );
   * console.log(data); // [4.2, 3.8, 4.5, 4.7, 3.9]
   */
  async floatArray(query: string, options?: MethodParamsOptions) {
    const data = await this.makeRequest<number[]>({
      endpointType: 'float_array',
      query,
      contextId: options?.contextId,
    });

    return data;
  }

  /**
   * More information: https://docs.gethyper.ai/types#boolean-array
   * @param {string} query - The query to be processed
   * @param {MethodParamsOptions} options - Optional parameters to be passed to the request, such as `contextId`
   * @returns The result of the query as a array of booleans
   * @example
   * const { data, error } = await hyper.types.booleanArray(
        'Are services meeting performance targets?',
        { contextId: '123e4567-e89b-12d3-a456-426614174000' },
     );
   * console.log(data); // [true, false, true, true, false]
   */
  async booleanArray(query: string, options?: MethodParamsOptions) {
    const data = await this.makeRequest<boolean[]>({
      endpointType: 'boolean_array',
      query,
      contextId: options?.contextId,
    });

    return data;
  }

  /**
   * More information: https://docs.gethyper.ai/types#datetime-array
   * @param {string} query - The query to be processed
   * @param {MethodParamsOptions} options - Optional parameters to be passed to the request, such as `contextId`
   * @returns The result of the query as a array of datetime strings
   * @example
   * const { data, error } = await hyper.types.datetimeArray(
        'What are the upcoming project deadlines?',
        { contextId: '123e4567-e89b-12d3-a456-426614174000' },
     );
   * console.log(data); // ['2023-11-15T17:00:00Z', '2023-12-01T17:00:00Z', '2023-12-20T17:00:00Z']
   */
  async datetimeArray(query: string, options?: MethodParamsOptions) {
    const data = await this.makeRequest<string[]>({
      endpointType: 'datetime_array',
      query,
      contextId: options?.contextId,
    });

    return data;
  }
}
